### DISCUSSION TITLE
How might you implement mocking (for unit testing) in a generic Forth?

### DISCUSSION THREAD

**Discussion by RickCarlino** [ForthHub/discussion#99] [Q&A]
One idea I have had is to do vectored execution for all words that trigger side effects.

For example, if you need to write to STDOUT (a side effect), you would wrap the `."` word in a second word that contains the XT of `."`. Under normal execution, your wrapper word would call the XT of `."`. In your test suite, you would change the value of the XT to point to a "dummy" version of `."` that does not actually write to STDOUT. That way, your test suite can run tests without actually performing the side effect.

Are there other ways to accomplish this?

NOTE: This is a [cross post of my discussion on /r/Forth (Reddit)](https://www.reddit.com/r/Forth/comments/nm71ac/how_might_you_implement_mocking_for_unit_testing/)

**Comment by iru-** [comment]
You could redefine the words you want in a testing harness and make your
test cases use (e.g. include) the harness.

Test cases would then reference mocked words, and all references would be
resolved at compile time instead of run time.

Le jeu. 27 mai 2021 à 19:38, Rick Carlino ***@***.***> a
écrit :

**Comment by massung** [comment]
There's a lot of difficulties in making something like this work in a "generic" Forth (whatever that means... maybe 200x?). For example:

```forth
: ." ... ;
: some-word ." foo" ;

' ." defer ." is ."

: test-."
  ['] dummy-." is ." some-word
  ;
```

When `test-."` is executed, `some-word` is still going to execute the original version of `."` on most (generic) Forths. In order for this to actually meet the intended goals, I'd think a Forth would need to have some version of Common Lisp's dynamically-scoped variables or at least the concept of an "output destination", be able to save/restore destinations as easily as it does input sources, and then side-effect words would need to written at the core to use them.

**Comment by GeraldWodni** [comment]
Most Forth systems I know already feature a vectored I/O. The easiest approach is to vector `emit`/`type` and `key?`/`key`. Some are [deferred](https://forth-standard.org/standard/core/DEFER) words.

__Gforth__
```forth
: xtype nip 0 ?do [char] x emit loop ; redefined xtype   ok         
' xtype is type
```

or if no side effect is wanted:
```forth
' 2drop is type
' drop is emit
```

You can always check the current state of a deferred word with [action-of](https://forth-standard.org/standard/core/ACTION-OF), that way you can make an `enter-silence` and `exit-silence` word-pair which respects the current vectoring when entering a silence-section and restores them at exit.

__VFX__
Has the generic `GENIO` approach, which aims to redirect more than the basics

__Mecrisp__
Has no separate vector for type, but vectors `emit`/`emit?` and `key`/`key?`

**Comment by cwpjr** [comment]
I know you should know that   ."  puts a return value for a completion of
it's function FVN (Forth Virtual Machine state) on the return stacke, vs
the data stack.
In my figgish I reset HERE to where it was before that word's compilation
bombed.

In forth you tend to your messes.
